runtime/src/accounts_db.rs

```diff
@@ -1958,51 +1963,56 @@ impl AccountsDb {
-                        accounts_index_map_lock_ref.unwrap(),
-                    )
-                } else {
-                    self.accounts_index.get_account_read_entry(pubkey)
-                };
+        for (pubkey, stored_account) in &stored_accounts {
+            let lookup = if is_startup {
+                self.accounts_index
+                    .get_account_read_entry_with_lock(pubkey, accounts_index_map_lock_ref.unwrap())
+            } else {
+                self.accounts_index.get_account_read_entry(pubkey)
+            };
 
-                if let Some(locked_entry) = lookup {
-                    let is_alive = locked_entry.slot_list().iter().any(|(_slot, i)| {
-                        i.store_id == stored_account.store_id
-                            && i.offset == stored_account.account.offset
-                    });
-                    if !is_alive {
-                        // This pubkey was found in the storage, but no longer exists in the index.
-                        // It would have had a ref to the storage from the initial store, but it will
-                        // not exist in the re-written slot. Unref it to keep the index consistent with
-                        // rewriting the storage entries.
-                        locked_entry.unref()
-                    } else {
-                        alive_total += stored_account.account_size as u64;
-                    }
-                    is_alive
+            if let Some(locked_entry) = lookup {
+                let is_alive = locked_entry.slot_list().iter().any(|(_slot, i)| {
+                    i.store_id == stored_account.store_id
+                        && i.offset == stored_account.account.offset
+                });
+                if !is_alive {
+                    // This pubkey was found in the storage, but no longer exists in the index.
+                    // It would have had a ref to the storage from the initial store, but it will
+                    // not exist in the re-written slot. Unref it to keep the index consistent with
+                    // rewriting the storage entries.
+                    unrefed_pubkeys.push(pubkey);
+                    locked_entry.unref();
                 } else {
-                    false
+                    alive_total += stored_account.account_size as u64;
+                    alive_accounts.push((pubkey, stored_account));
                 }
-            })
-            .collect();
+            }
+        }
         drop(accounts_index_map_lock);
         index_read_elapsed.stop();
         let aligned_total: u64 = self.page_align(alive_total);
 
+        // If not saving memory maps or more than 1 page of data, then skip the shrink
+        if num_stores == 1 && aligned_total + PAGE_SIZE >= original_bytes {
+            for pubkey in unrefed_pubkeys {
+                if let Some(locked_entry) = self.accounts_index.get_account_read_entry(pubkey) {
+                    locked_entry.addref();
```

Comment:
I wonder if we could avoid this `addref()` by either:

1) Removing the earlier `unref()`, and instead just  iterate through the `unrefed_pubkeys` and call `unref()` after we've passed the `aligned_total + PAGE_SIZE >= original_bytes` check above. Downside is an extra pass through the `unrefed_pubkeys` keys

2) We have a `store.alive_bytes()` and `store.total_bytes()`. Maybe using these we could move the `aligned_total + PAGE_SIZE >= original_bytes` check earlier in the function before the scan, and exit there if the check fails (would also avoid the overhead of the additional scan through the storages).  Something like:

`if num_stores == 1 && self.page_align(store.alive_bytes() as u64) + PAGE_SIZE >= store.total_bytes() { return 0;}`

Reply:
Right for 1, I am thinking this is the not-common case since we queued it up to shrink and we have other signals that tell us that shrinking is likely productive like alive_bytes/written_bytes. So we should optimize for the case where we are actually doing the shrink.

Clean does run before this shrink. I was thinking those may not be accurate, hence the location of this check, but I think you might be right, they might actually be somewhat accurate.